backend/notifications/services.py

```
@@ -0,0 +1,515 @@
+def internal_route(path):
+    """Normalize an internal SPA path to the hash route the frontend uses."""
+    if not path:
+        return ''
+    if path.startswith('#/') or path.startswith('http://') or path.startswith('https://'):
+        return path
+    if not path.startswith('/'):
+        path = f'/{path}'
+    return f'#{path}'
+
+
+def _source_for(obj):
```

Ruff (0.15.15)

[warning] Missing return type annotation for private function `_source_for`

(ANN202)